designs/2024-per-rule-autofix-configuration/README.md

```diff
@@ -0,0 +1,406 @@
+* [types/index.d.ts > RuleEntry](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/types/index.d.ts#L1377)
+* [types/index.d.ts > LinterOptions](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/types/index.d.ts#L1876)
+* ---
+* [flat-config-schema.js > rulesSchema](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/config/flat-config-schema.js#L450)
+* [flat-config-schema.js > disableDirectiveSeveritySchema](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/config/flat-config-schema.js#L298)
+* [flat-config-schema.js > normalizeRuleOptions](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/config/flat-config-schema.js#L132)
+* [config.js > #normalizeRulesConfig](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/config/config.js#L515)
+* [config.js > validateRulesConfig](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/config/config.js#L553)
+* [linter/apply-disable-directives.js](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/linter/apply-disable-directives.js#L474) should take a full form of the entry.
+
+Some of static methods / individual functions also need to be considered, 
+like [config.js > getRuleNumericSeverity](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/config/config.js#L634)
+or [linter.js > getRuleOptions](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/linter/linter.js#L1086).
+
+There are likely additional functions that operate on non-normalized rule entries; tests should help identify them. \
+There's also an open question here regarding support for the legacy config format, as the current code relies on `@eslint/eslintrc` in some places.
+
+Next, the CLI engine and directive parser should be verified to support object-form rule entries.
+
+---
+
+Next, all consumers of normalized rule entries and `reportUnusedDisableDirectives` must be updated accordingly.
+
+For directives, updating [apply-disable-directives.js](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/linter/apply-disable-directives.js#L466) may be sufficient.
+
+For rules usage is more widespread, there are a lot of usage like [this](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/linter/linter.js#L699), [here](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/linter/linter.js#L2140) and [there](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/linter/linter.js#L1086). \
+I'm not sure it's possible to identify them all and that it's necessary to list them here, hopefully the tests will help.
+
+Once normalization is in place and consistent, we can safely introduce autofix disabling logic.
```

Comment:
Is the recommendation here to update rule config normalization in a separate PR, before introducing autofix disabling logic? Can this be clarified?